### PR TITLE
Record compilation time in benchmarks and allow plotting benchmarks from multiple runs

### PR DESCRIPTION
**Pull request by matt-graham**
This PR adds some functionality to the benchmarks suite:

- An additional `compilation_time_in_seconds` value is recorded in the benchmark results, corresponding to the time taken in seconds to run ahead-of-time-compilation on any JAX benchmark functions which are marked to be JIT compiled using the `jit_benchmark` flag passed back by the benchmark `setup` function. We were already separately performing this ahead of time compilation to avoid it biasing the run time results, but timing this gives useful information about code optimizations that improve compilation time performance like #290. 
- Rather than closing the benchmark function over all arguments passing the argumentless function to the JAX `jit` transform, the arguments corresponding to the benchmark parameter set are instead explicitly passed in when compiling / running the function, and any array arguments in these converted to `jax.Array` objects. The conversion step should hopefully cause any host-to-device transfers to happen outside of the benchmark run time evaluation and so give a better reflection of performance when data is already on device (which is the more relevant figure for comparison). Not closing over all arguments guards against the (speculated) possibility of the XLA compiler recognizing that it can [constant propagate](https://en.wikipedia.org/wiki/Constant_folding) through the full computation as all values are known at run time.
- The benchmark plotting module is updated to allow overlaying the benchmark results from multiple independent runs in different outputted JSON files on the same plots, to allow for example more easily comparing benchmarks before and after code changes or across different devices. The plotting functions are also refactored as part of this to avoid some of the redundancy in the code.

### CODE REVIEW

File: benchmarks/benchmarking.py [hunk 1]
```diff
@@ -369,11 +374,36 @@ def trace_memory_allocations(n_frames: int = 1) -> Callable[[], tuple[int, int]]
...
+    args.update(
+        jax.tree.map(
+            lambda x: jax.numpy.asarray(x) if isinstance(x, np.ndarray) else x, args
+        )
+    )
+    with timer() as t:
+        compiled_benchmark_function = (
+            jax.jit(benchmark_function).lower(**args).compile()
+        )
+    results_entry["compilation_time_in_seconds"] = t()
```
Comment (willGraham01) [inline benchmarks/benchmarking.py]:
```suggestion
    with timer() as compilation_timer:
        compiled_benchmark_function = (
            jax.jit(benchmark_function).lower(**args).compile()
        )
    results_entry["compilation_time_in_seconds"] = compilation_timer()
```

File: benchmarks/benchmarking.py [hunk 2]
```diff
@@ -369,11 +374,36 @@ def trace_memory_allocations(n_frames: int = 1) -> Callable[[], tuple[int, int]]
...
+    Based on StackOverflow answer https://stackoverflow.com/a/69156219 by Justin Dehorty
+    """
+    time_0 = time_1 = timeit.default_timer()
+    yield lambda: time_1 - time_0
+    time_1 = timeit.default_timer()
+
+
 def _compile_jax_benchmark_and_analyse(
-    benchmark_function: Callable, results_entry: dict
+    benchmark_function: Callable, args: dict, results_entry: dict
```
[the remark this answers is not in the document]
Comment (matt-graham) [reply benchmarks/benchmarking.py]: To clarify as I realised on making changes I do 'splat' this as `**args` what I  said above is very consuing - I meant not capturing as keyword arguments in function signature as `**kwargs`. I think overall `kwargs` is probably the better name here though so have updated to that to follow your suggestion.

File: benchmarks/plotting.py [hunk 3]
```diff
@@ -42,96 +53,94 @@ def _plot_scaling_guide(
...
 
-def plot_results_against_bandlimit(
-    benchmark_results_path: str | Path,
-    functions: tuple[str] = ("forward", "inverse"),
-    measurements: tuple[str] = ("times", "flops", "memory", "error"),
-    axis_size: float = 3.0,
+def plot_results_against_parameter(
+    benchmark_results_paths: list[str | Path],
+    functions: tuple[str],
+    metric_names: tuple[str],
```
[the remark this answers is not in the document]
Comment (matt-graham) [reply benchmarks/plotting.py]: Defaults now updated to be closer to what they were previously though have maintained swapping out plotting FLOP count for compilation time by default, as I think the latter is the more useful metric to include by default.
